packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -72,88 +81,145 @@ export function endTrackFrame(): Tag {
-  context?: string | false
-): (a1: U, a2: V, a3: W) => T;
-export function memo<T, U, V, W, X>(
-  cb: (a1: U, a2: V, a3: W, a4: X) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W, a4: X) => T;
-export function memo<T, U, V, W, X, Y>(
-  cb: (a1: U, a2: V, a3: W, a4: X, a5: Y) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W, a4: X, a5: Y) => T;
-export function memo<T, U, V, W, X, Y, Z>(
-  cb: (a1: U, a2: V, a3: W, a4: X, a5: Y, a6: Z) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W, a4: X, a5: Y, a6: Z) => T;
-
-export function memo<T>(callback: (...args: unknown[]) => T, debuggingContext?: string | false) {
-  let lastValue: T | undefined;
-  let tag: Tag;
-  let snapshot: Revision;
-
-  let memoized = (...args: unknown[]): T => {
-    if (!tag || !validateTag(tag, snapshot)) {
-      beginTrackFrame();
-
-      try {
-        if (DEBUG) {
-          runInAutotrackingTransaction!(() => (lastValue = callback(...args)), debuggingContext);
-        } else {
-          lastValue = callback(...args);
-        }
-      } finally {
-        tag = endTrackFrame();
-        snapshot = valueForTag(tag);
-        consumeTag(tag);
-
-        // If the final tag is constant, then we know for sure that this
-        // memoized function can never change. There are times when this
-        // information is useful externally (i.e. in the append VM, it tells us
-        // whether or not to emit opcodes) so we expose it via a metadata weakmap.
-        if (tag === CONSTANT_TAG) {
-          IS_CONST_MAP.set(memoized, true);
-        } else if (DEBUG) {
-          // In DEBUG, set the value to false explicitly. This way we can throw
-          // if someone attempts to call `isConst(memoized)` before running
-          // `memoized()` at least once.
-          IS_CONST_MAP.set(memoized, false);
-        }
-      }
-    } else {
-      consumeTag(tag);
-    }
+const CACHE_KEY: unique symbol = symbol('CACHE_KEY');
+
+export function memo<T>(callback: () => T, debuggingContext?: string | false) {
+  let cache = createCache(callback, debuggingContext);
+
+  let memoized = () => getValue(cache);
+
+  (memoized as any)[CACHE_KEY] = cache;
+
+  return memoized;
+}
+
+export function isConstMemo(memoized: Function) {
+  let cache = (memoized as any)[CACHE_KEY] as Cache | undefined;
+
+  return cache === undefined ? false : isConst(cache);
+}
 
-    return lastValue!;
+//////////
+
+// public interface
+export interface Cache<T = unknown> {
+  [CACHE_KEY]: T;
+}
+
+const FN: unique symbol = symbol('FN');
+const LAST_VALUE: unique symbol = symbol('LAST_VALUE');
+const TAG: unique symbol = symbol('TAG');
+const SNAPSHOT: unique symbol = symbol('SNAPSHOT');
+const DEBUG_LABEL: unique symbol = symbol('DEBUG_LABEL');
+
+interface InternalCache<T = unknown> {
+  [FN]: (...args: unknown[]) => T;
+  [LAST_VALUE]: T | undefined;
+  [TAG]: Tag | undefined;
+  [SNAPSHOT]: Revision;
+  [DEBUG_LABEL]?: string | false;
```

Comment:
Just to make sure, the `false` here is for the `DEBUG && somethingPassedIn` cases?

Reply:
yep, exactly. We can revisit that in the future, it's just convenient for the time being